action panel reimplementation

wlitwin opened the pull request:
WIP pull request so it's easy to keep tabs on my progress, and so it's easy to provide feedback. So far I've only done the minimum to get a side bar up and enabling/disabling buttons based on edit state. There's a number of styling/formatting issues with the left panel. Notably closing is broken at the moment. Also, some of my descriptions are probably inaccurate, and there's some informational nodes missing.

![image](https://user-images.githubusercontent.com/1189464/75642683-343ed380-5bf1-11ea-99a7-2367617766f2.png)

wlitwin committed:
Bring back ActionPanel.re and get it to show

Still broken, but shows up on the web page now. Had to add a few
methods back to web/Model.re. Seems the HTML/CSS is a bit broken
and doesn't behave quite like the right panel. Also, semantics/Action.re
has changed quite a bit since the panel was last used.

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin committed:
Start implementing a new action panel, inspired by the old panel

cyrus- commented:
Let's maybe have it be a little more grouped this time, rather than just a vertical list.

So, for example, maybe have:

Movement:
- move using arrow keys
- click to move directly to a location
- move to next/prev hole: tab/shift+tab

Arithmetic:
- enter number literals directly
- arithmetic operators: + - *

Variables:
- enter variables directly
- type "let<space>" to insert a let expression

(and so on for the other features)

wlitwin committed:
Add sub-sections for the actions panel

wlitwin committed:
Fix description and add new category for injections

wlitwin committed:
Kind of fix the left pane collapse

Unfortunately the collapse animation is broken for the inner action
pane. Tried a few things but not sure what's preventing it from
happening.

wlitwin committed:
Add a section for types and allow enable/disabling of info elements

wlitwin commented:
![image](https://user-images.githubusercontent.com/1189464/76042657-3ca55000-5f0a-11ea-850d-accba11aceb7.png)

New screenshot with sections

cyrus- commented:
Looks good!

Small suggestions:
* Change 'type "let"' to 'type "let<space>"' to emphasize that you need to press space
* Make things the user needs to type ("let", "true", "false", etc.) use the mono font that we use for code
* Change "Insert a list" to "Empty list (nil)"
* Add a "functions" section and put lambda and function application (space) there
* "a numeric type" -> "the Num type"
* "a boolean type" -> "the Bool type"
* Put the type ascription thing under both let and functions, and rename the "type ascription" section to "types"
* The "|" operator is part of types, so put that under types too

Also, let's experiment with making better use of horizontal space? For example:
* move to prev/next hole could maybe be on one line, with two hotkeys shown
* the arithmetic operators could all be summarized on one line

Feel free to experiment with other ways to summarize things.

Thanks!

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

cyrus- commented:
@wlitwin how's it going?

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin committed:
Change the DSL definition to use plain functions

There's much more complex layout so the type DSL is now a function DSL.
Attempted to combine some of the operators. Need to make a more general
way to combine different things, a fancier "action button".

wlitwin committed:
Fix bug with operators that require special cursor locations

The operator > is used for boolean > and type ascription. It's listed
in two places on the action panel and they're each sensitive to
different cursor contexts. This commit checks the cursor context before
asking if the action is allowed.

wlitwin committed:
Experiment with putting some actions on the same line

wlitwin commented:
Still thinking of different ways to combine things
![Selection_001](https://user-images.githubusercontent.com/1189464/81247306-7e1baf80-8fce-11ea-8dba-264bfa3fcb6a.png)

cyrus- commented:
Looking good! Delete and Backspace do the same thing in different directions, so they should probably be grouped together. Also type ascription is in there twice. 

I assume you have a section on functions below the fold?

cyrus- commented:
Also the latest commits to dev by @Yunif3 added floating point numbers and dotted float ops like in OCaml.

PPER commented:
> Looking good! Delete and Backspace do the same thing in different directions, so they should probably be grouped together. Also type ascription is in there twice.
> 
> I assume you have a section on functions below the fold?

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin commented:
> I assume you have a section on functions below the fold?

Yes! There's also a miscellaneous section that has some commands that I'm not sure where to place.

![Selection_003](https://user-images.githubusercontent.com/1189464/81459610-c6b1a500-9155-11ea-8549-feb6f9989699.png)

> Also the latest commits to dev by @Yunif3 added floating point numbers and dotted float ops like in OCaml.

Cool! I'll add a note about those

wlitwin committed:
Group delete + backspace, remove duplicate type ascription

cyrus- commented:
* "Separate terms" should be "apply function"
* "Parenthesize expression" works for patterns and types too, so just call it "Parenthesize"
* the "insert , operator" one should be in a section called "tuples"
* "insert case expression" is now entered with the "case" keyword, similar to "let" (though maybe Alt+C also still works?). it should be in a section called "Pattern Matching". you might also want to write "add new rule" in that section, with "Enter" as the key.

wlitwin committed:
Make requested changes

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin committed:
Add note about floating point operators

wlitwin commented:
>though maybe Alt+C also still works?
Seems to still work! I left it in, unless this is planning to be phased out?

Latest changes

![Screen Shot 2020-05-10 at 14 53 18](https://user-images.githubusercontent.com/1189464/81511414-259a2a00-92ce-11ea-923d-d53b6345b92d.png)

cyrus- commented:
- The Delete key does not "delete expression", it just does the same thing as backspace but to the right of the cursor. Maybe just have an entry that says "Backspace / Delete" and then has the two keys on the right.
- Everything from Backspace on down in the "Movement" section is not about movement, but rather general editing. Let's create a new section called "General Editing" that includes that + the "Parenthesize" action that is now at the bottom.
- The floating point operator characters can just be in the light brown color on the right, just like the integer operators.
- We don't have a Num type anymore, we have Int and Float ("I" and "F", respectively -- don't need to write "Shift+".
- Let's put the type construction into the corresponding semantic sections, instead of in the "Types" section.
- For variables, maybe we can list the regex for valid variables?
- "injunction" should be "injection". call that section "sum types"
- "type arrow" should be "arrow type"

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin committed:
Address pull request comments

 - The Delete key does not "delete expression", it just does the same thing as backspace but to the right of the cursor. Maybe just have an entry that says "Backspace / Delete" and then has the two keys on the right.

 - Everything from Backspace on down in the "Movement" section is not about movement, but rather general editing. Let's create a new section called "General Editing" that includes that + the "Parenthesize" action that is now at the bottom.

 - The floating point operator characters can just be in the light brown color on the right, just like the integer operators.

 - We don't have a Num type anymore, we have Int and Float ("I" and "F", respectively -- don't need to write "Shift+".

 - Let's put the type construction into the corresponding semantic sections, instead of in the "Types" section.

 - For variables, maybe we can list the regex for valid variables?

 - "injunction" should be "injection". call that section "sum types"

 - "type arrow" should be "arrow type"

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin commented:
Latest changes applied:

![Screen Shot 2020-05-25 at 22 07 19](https://user-images.githubusercontent.com/1189464/82862343-3e146200-9ed4-11ea-92c5-082ef8cea22f.png)

cyrus- commented:
@wlitwin can you push your code changes? it is probably easier for me to make comments directly on the code for the various messages

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+            ],
+          ),
+        ]),
+        combo(Ctrl_Alt_I, simple("Swap expression up")),
```
* this should be "swap line up/down"
* can we combine the up/down ones into a dual_line?

wlitwin replied:
I changed it to a the same style as `Backspace / Delete` as it looked a bit weird with the action boxes interspersed with the text. Now the whole UI is more uniform.

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+        ]),
+        combo(Ctrl_Alt_I, simple("Swap expression up")),
+        combo(Ctrl_Alt_K, simple("Swap expression down")),
+        combo(Ctrl_Alt_J, simple("Swap expression left")),
```
this should be "swap operand left/right"

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+          Action.Construct(SChar("F")),
+        ),
+        spaced_line([
+          text("Floating point operators, type"),
```
drop the ", type"

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+        ]),
+        operator_list(
+          ~on_type=false,
+          "Integer operators",
```
Put integer operators first, to be consistent with order of the type insertion lines

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+        ),
+      ],
+    ),
+    section(
```
lets move this section above arithmetic

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+      ],
+    ),
+    section(
+      "Types",
```
lets get rid of this section and move all of this stuff into the sections for the corresponding expression operations

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+    section(
+      "Types",
+      [
+        combo(Colon, simple("Type ascription")),
```
move to variables section

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+      "Types",
+      [
+        combo(Colon, simple("Type ascription")),
+        combo(VBar, simple("Insert | operator")),
```
move to sum types section

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+        combo_and_cursor(
+          ~on_type=true,
+          LeftBracket,
+          simple("Insert type List"),
```
move to lists section

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+          LeftBracket,
+          simple("Insert type List"),
+        ),
+        combo_and_cursor(~on_type=true, GT, [text("Create an arrow type")]),
```
move to functions section

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,464 @@
+        ),
+        operator_list(
+          ~on_type=false,
+          "Operators",
```
these are actually integer comparison operators only. there are also dotted versions of these for float comparison. let's include these in the arithmetic section, and move the boolean section above it so that booleans are described before the comparison operators.

wlitwin committed:
Address PR review comments

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin committed:
line -> operand

wlitwin commented:
![Screen Shot 2020-05-27 at 12 03 16](https://user-images.githubusercontent.com/1189464/83061585-1d522680-a012-11ea-839c-7a71dc183136.png)

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+    section(
+      "Variables",
+      [
+        info([text("Variable regex: ^[_a-z][_a-zA-Z0-9']*$")]),
```
we can drop the `^` and `$`. also the latest `dev` branch allows initial caps.

cyrus- replied:
also can you use code font for the regex

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+      [
+        info([
+          text("Enter boolean literals "),
+          mono_text("\"true\", \"false\""),
```
don't need quotation marks around true and false

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+          ],
+          Action.Construct(SChar("B")),
+        ),
+        operator_list(~on_type=false, "Logical AND", [Ampersand]),
```
also add OR (`|`) to this entry

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+        combo_and_cursor(
+          ~on_type=false,
+          LeftBracket,
+          simple("Insert Empty List (nil)"),
```
replace "Empty List" with `[]` (in code font)

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+          [
+            text("Type "),
+            mono_text("\"I\""),
+            text(" to insert an Integer type"),
```
replace "Integer" with `Int` (in code font)

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+          [
+            text("Type "),
+            mono_text("\"F\""),
+            text(" to insert a Float type"),
```
put `Float` in code font

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+        ),
+        operator_list(
+          ~on_type=false,
+          "Integer operators",
```
latest `dev` has division too `/` and `/.`

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+      [
+        combo(Alt_L, simple("Left injection")),
+        combo(Alt_R, simple("Right injection")),
+        combo(VBar, simple("Insert | operator")),
```
replace "operator" with "type operator" (this is currently broken on `dev`)

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,437 @@
+    section(
+      "Functions",
+      [
+        combo(Backslash, simple("Insert Lambda expression")),
```
replace "Lambda expression" with "lambda function"

cyrus- commented:
Looks pretty close to ready! Another thing we should do is add a simple "checkin" function, the purpose of which is to alert anyone adding a new action that they need to modify this sidebar too.

```
type ack_checkin = AddedAction

/* add a new case to this function only if you have already added that action to the sidebar! */
let checkin = (a : Action.t) => 
  switch (a) {
  | Construct(SVar(_)) -> AddedAction
  | etc
```

wlitwin committed:
Address pull request comments

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin committed:
Fix merge breaks

wlitwin commented:
![Screen Shot 2020-05-28 at 13 43 41](https://user-images.githubusercontent.com/1189464/83191697-48f41000-a0e9-11ea-9b48-8e99e15fcfa6.png)

wlitwin committed:
Move left/right are mentioned, "use arrow keys"

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,501 @@
+        combo_and_cursor(~on_type=true, GT, [text("Create an arrow type")]),
+      ],
+    ),
+    section("Tuples", [combo(Comma, simple("Create a tuple"))]),
```
Let's add another entry for create a tuple type, which also has Comma

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,501 @@
+  | SwapLeft
+  | SwapRight
+  | MoveLeft
+  | MoveRight => ()
```
let's define a type with one constructor, `Added`, instead of using `()`, to avoid people just reflexively copying. also let's do one switch rule per action, instead of or-ing them together like this.

cyrus- commented:
don't forget to click "Ready for review" on the PR when the changes are ready.

wlitwin committed:
Merge remote-tracking branch 'origin/dev' into dev

wlitwin committed:
Address review comments

wlitwin commented:
![Screen Shot 2020-05-30 at 16 32 20](https://user-images.githubusercontent.com/1189464/83340979-361d3f00-a293-11ea-9f15-51ce3d39e784.png)

cyrus- commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -0,0 +1,514 @@
+  | MoveLeft => Added
+  | MoveRight => Added
+  /* Not added */
+  | Construct(SApPalette(_)) => Added
```
change the Palette-related ones to `failwith("Unimplemented")`

wlitwin committed:
Change the unused cases to failwith("Unimplemented")

wlitwin committed:
Oops, palette only ones to failwith

cyrus- commented:
looks good!
